framework/src/positions/Positions.C

```diff
@@ -0,0 +1,244 @@
+    mooseError("Inconsistency between the 2D and 1D position vectors detected. "
+               "The 2D positions must unroll into the 1D positions");
+
+  // 3D vs 2D & 1D
+  if (temp_3d_unrolled.size() && temp_2d_unrolled.size() && temp_3d_unrolled != temp_3d_unrolled)
+    mooseError("Inconsistency between the 3D and 2D position vectors detected. "
+               "The 3D positions must unroll the same way as the 2D positions");
+  if (temp_3d_unrolled.size() && _positions.size() && temp_3d_unrolled != _positions)
+    mooseError("Inconsistency between the 3D and 1D position vectors detected. "
+               "The 3D positions must unroll into the 1D positions");
+
+  // 4D vs all lower dimensions
+  if (temp_4d_unrolled.size() && temp_3d_unrolled.size() && temp_4d_unrolled != temp_3d_unrolled)
+    mooseError("Inconsistency between the 4D and 3D position vectors detected. "
+               "The 4D positions must unroll the same way as the 3D positions");
+  if (temp_4d_unrolled.size() && temp_2d_unrolled.size() && temp_4d_unrolled != temp_2d_unrolled)
+    mooseError("Inconsistency between the 4D and 2D position vectors detected. "
+               "The 4D positions must unroll the same way as the 2D positions");
+  if (temp_4d_unrolled.size() && _positions.size() && temp_4d_unrolled != _positions)
+    mooseError("Inconsistency between the 4D and 1D position vectors detected. "
+               "The 4D positions must unroll into the 1D positions");
+
+  // Use any of the higher D to set the one D vector
+  if (!_positions.size())
+  {
+    if (temp_2d_unrolled.size())
+      _positions = temp_2d_unrolled;
+    else if (temp_3d_unrolled.size())
+      _positions = temp_3d_unrolled;
+    else if (temp_4d_unrolled.size())
+      _positions = temp_4d_unrolled;
+    else
+      mooseError("Positions::unrollMultiDPositions() may only be called if there is at least one "
+                 "non-empty positions vector.");
+  }
+}
+
+void
+Positions::finalize()
+{
+  // Gather up the positions vector on all ranks
+
+  if (_need_broadcast)
+  {
+    // TODO: be able to do this
+    // _communicator.allgather(_positions, /* identical buffer lengths = */ false);
+
+    // Other notes to explain why we do this:
+    // - The vector broadcasting code is for VPPs, so it's *currently* not accessible as we have
+    //   vectors of points, not reals
+    // - The consumer/producer reporter interface can keep track of whether a reduction is needed
+    //   (for example if a consumer needs replicated data, but the producer is distributed) however,
+    //   we have currently made the decision that positions should ALWAYS be replicated
+
+    // Gather all components of all positions
+    std::vector<Real> x(_positions.size()), y(_positions.size()), z(_positions.size());
+    for (const auto & pt : _positions)
+    {
+      x[&pt - &_positions[0]] = pt(0);
+      y[&pt - &_positions[0]] = pt(1);
+      z[&pt - &_positions[0]] = pt(2);
+    }
```

Comment:
I think the idiom where you loop over indexes is clearer to read

Reply:
will change if cant use parallel algebra

Reply:
Removed those loops